test/core/storage/trie/trie_db/trie_batch_test.cpp

```diff
@@ -0,0 +1,191 @@
+#include "crypto/hasher/hasher_impl.hpp"
+#include "testutil/literals.hpp"
+#include "testutil/outcome.hpp"
+#include "testutil/storage/base_leveldb_test.hpp"
+#include "testutil/storage/polkadot_trie_db_printer.hpp"
+
+using namespace kagome::storage::trie;
+using kagome::storage::face::WriteBatch;
+using kagome::common::Buffer;
+using kagome::hash::HasherImpl;
+using testing::_;
+using testing::Invoke;
+using testing::Return;
+
+class TrieBatchTest : public test::BaseLevelDB_Test {
+ public:
+  TrieBatchTest() : BaseLevelDB_Test("/tmp/leveldbtest") {}
+
+  void SetUp() override {
+    open();
+    trie = std::make_unique<PolkadotTrieDb>(std::move(db_));
+  }
+
+  static const std::vector<std::pair<Buffer, Buffer>> data;
+
+  std::unique_ptr<PolkadotTrieDb> trie;
+};
+
+const std::vector<std::pair<Buffer, Buffer>> TrieBatchTest::data = {
+    {"123456"_hex2buf, "42"_hex2buf},
+    {"1234"_hex2buf, "1234"_hex2buf},
+    {"010203"_hex2buf, "0a0b"_hex2buf},
+    {"010a0b"_hex2buf, "1337"_hex2buf},
+    {"0a0b0c"_hex2buf, "deadbeef"_hex2buf}};
+
+void FillSmallTrieWithBatch(WriteBatch<Buffer, Buffer>& batch) {
+  for (auto &entry : TrieBatchTest::data) {
+    EXPECT_OUTCOME_TRUE_void(_, batch.put(entry.first, entry.second));
+  }
+}
+
+class MockPolkadotTrieDb : public PolkadotTrieDb {
+ public:
+  MockPolkadotTrieDb() : PolkadotTrieDb(nullptr) {}
+
+  MOCK_CONST_METHOD0(getRootHash, Buffer());
+};
+
+class MockDb : public test::MapDb {
+ public:
+  MOCK_METHOD2(put, outcome::result<void>(const Buffer &, const Buffer &));
+
+  // to retain the ability to call the actual implementation of put from the
+  // superclass
+  outcome::result<void> true_put(const Buffer &key, const Buffer &value) {
+    return MapDb::put(key, value);
+  }
+};
+
+/**
+ * @given an empty trie
+ * @when putting some entries into it using a batch
+ * @then all inserted entries are accessible from the trie
+ */
+TEST_F(TrieBatchTest, Put) {
+  auto batch = trie->batch();
+  FillSmallTrieWithBatch(*batch);
+
+  for (auto &entry : data) {
+    EXPECT_OUTCOME_TRUE(res, trie->get(entry.first));
+    ASSERT_TRUE(res.empty());
+  }
+  EXPECT_OUTCOME_TRUE_void(r, batch->commit());
+  for (auto &entry : data) {
+    EXPECT_OUTCOME_TRUE(res, trie->get(entry.first));
+    ASSERT_EQ(res, entry.second);
+  }
+
+  EXPECT_OUTCOME_TRUE_void(__, trie->put("102030"_hex2buf, "010203"_hex2buf));
+  EXPECT_OUTCOME_TRUE_void(_, trie->put("104050"_hex2buf, "0a0b0c"_hex2buf));
+  EXPECT_OUTCOME_TRUE(v1, trie->get("102030"_hex2buf));
+  ASSERT_EQ(v1, "010203"_hex2buf);
+  EXPECT_OUTCOME_TRUE(v2, trie->get("104050"_hex2buf));
+  ASSERT_EQ(v2, "0a0b0c"_hex2buf);
+}
+
+/**
+ * @given a small trie
+ * @when removing some entries from it using a batch
+ * @then removed entries are no longer in the trie, while the rest of them stays
+ */
+TEST_F(TrieBatchTest, Remove) {
+  auto batch = trie->batch();
+  FillSmallTrieWithBatch(*batch);
+
+  EXPECT_OUTCOME_TRUE_void(_, batch->remove(data[2].first));
```

Comment:
strange sequence of calls. can you explain?
why not write a cycle over items?

Reply:
I just need to remove several items from known data set, and as their quantity is small, a loop wouldn't even decrease the number of code lines here.

Reply:
```for (auto i : {2, 3, 4}) {
  ASSERT_TRUE(batch->remove(data[i].first));
}```
The same line count, but looks better in my opinion.
Though, I don't think you need EXPECT_OUTCOME_TRUE_... here, ASSERT_TRUE is enough since you are not making use of result value.

Reply:
ok.